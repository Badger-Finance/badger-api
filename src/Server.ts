import { Configuration, Inject, PlatformApplication } from "@tsed/common";
import { controllers } from "./ControllerRegistry";
import methodOverride from "method-override";
import cookieParser from "cookie-parser";
import bodyParser from "body-parser";
import compress from "compression";

const rootDir = __dirname;

@Configuration({
  rootDir,
  acceptMimes: ["application/json"],
  mount: {
    "/v2/": controllers
  },
})
export class Server {
  @Inject()
  app!: PlatformApplication;

  /**
   * This method let you configure the express middleware required by your application to works.
   * @returns {Server}
   */
  public $beforeRoutesInit(): void | Promise<any> {
    this.app
      .use(cookieParser())
      .use(compress({}))
      .use(methodOverride())
      .use(bodyParser.json())
      .use(bodyParser.urlencoded({
        extended: true
      }));
  }
}
