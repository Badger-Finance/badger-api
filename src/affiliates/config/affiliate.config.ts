import { BadRequest } from '@tsed/exceptions';
import { Chain } from '../../chains/config/chain.config';
import { Protocol } from '../../config/constants';
import { SettAffiliateData } from '../../setts/interfaces/sett-affiliate-data.interface';
import { SettDefinition } from '../../setts/interfaces/sett-definition.interface';

type Affiliates = { [protocol: string]: Affiliate };

export abstract class Affiliate {
  private static affiliates: Affiliates = {};

  readonly name: string;

  constructor(name: string) {
    this.name = name;
  }

  static register(protocol: Protocol, affiliate: Affiliate): void {
    this.affiliates[protocol] = affiliate;
  }

  static getAffiliate(affiliateData: SettAffiliateData): Affiliate {
    const affiliate = this.affiliates[affiliateData.protocol];
    if (!affiliate) {
      throw new BadRequest(`${affiliate} is not a affiliate`);
    }
    return affiliate;
  }

  abstract getAffiliateVaultData(chain: Chain, sett: SettDefinition): Promise<SettAffiliateData>;
}
