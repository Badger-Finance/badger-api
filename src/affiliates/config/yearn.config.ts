import { BigNumber, ethers } from 'ethers';
import { Chain } from '../../chains/config/chain.config';
import { guestListAbi } from '../../config/abi/guest-list.abi';
import { yearnAffiliateVaultWrapperAbi } from '../../config/abi/yearn-affiliate-vault-wrapper.abi';
import { Protocol } from '../../config/constants';
import { SettAffiliateData } from '../../setts/interfaces/sett-affiliate-data.interface';
import { SettDefinition } from '../../setts/interfaces/sett-definition.interface';
import { getToken } from '../../tokens/tokens.utils';
import { Affiliate } from './affiliate.config';

export class Yearn extends Affiliate {
  constructor() {
    super(Protocol.Yearn);
    Affiliate.register(Protocol.Yearn, this);
  }

  async getAffiliateVaultData(chain: Chain, sett: SettDefinition): Promise<SettAffiliateData> {
    const affiliateData: SettAffiliateData = {
      protocol: Protocol.Yearn,
    };
    if (sett.hasBouncer) {
      try {
        const vaultWrapper = new ethers.Contract(sett.settToken, yearnAffiliateVaultWrapperAbi, chain.provider);
        const guestListAddress = await vaultWrapper.guestList();
        const guesList = new ethers.Contract(guestListAddress, guestListAbi, chain.provider);
        const [available, limit]: [BigNumber, BigNumber] = await Promise.all([
          guesList.remainingTotalDepositAllowed(),
          guesList.totalDepositCap(),
        ]);
        const depositToken = getToken(sett.depositToken);
        affiliateData.availableDepositLimit = parseFloat(ethers.utils.formatUnits(available, depositToken.decimals));
        affiliateData.depositLimit = parseFloat(ethers.utils.formatUnits(limit, depositToken.decimals));
      } catch {
        affiliateData.availableDepositLimit = 0;
        affiliateData.depositLimit = 50;
      }
    }
    return affiliateData;
  }
}
