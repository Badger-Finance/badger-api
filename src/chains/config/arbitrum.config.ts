import { Protocol } from '../../config/enums/protocol.enum';
import { SettState } from '../../config/enums/sett-state.enum';
import { Stage } from '../../config/enums/stage.enum';
import rpc from '../../config/rpc.config';
import { TOKENS } from '../../config/tokens.config';
import { GasPrices } from '../../gas/interfaces/gas-prices.interface';
import { getCurveSettTokenBalance } from '../../protocols/strategies/convex.strategy';
import { BouncerType } from '../../rewards/enums/bouncer-type.enum';
import { SettDefinition } from '../../setts/interfaces/sett-definition.interface';
import { arbitrumTokensConfig } from '../../tokens/config/arbitrum-tokens.config';
import { ChainNetwork } from '../enums/chain-network.enum';
import { ArbitrumStrategy } from '../strategies/arbitrum.strategy';
import { Chain } from './chain.config';

export class Arbitrum extends Chain {
  constructor() {
    super(
      'Arbitrum',
      'arbitrum',
      '0xa4b1',
      ChainNetwork.Arbitrum,
      arbitrumTokensConfig,
      arbitrumSetts,
      rpc[ChainNetwork.Arbitrum],
      new ArbitrumStrategy(Object.keys(arbitrumTokensConfig)),
      2425847,
      '0x635EB2C39C75954bb53Ebc011BDC6AfAAcE115A6',
      '0x85E1cACAe9a63429394d68Db59E14af74143c61c',
    );
    Chain.register(this.network, this);
  }

  async getGasPrices(): Promise<GasPrices> {
    const gasPrice = 2;
    return {
      rapid: gasPrice,
      fast: gasPrice,
      standard: gasPrice,
      slow: gasPrice,
    };
  }
}

export const arbitrumSetts: SettDefinition[] = [
  {
    name: 'Sushiswap Wrapped Ether/Sushi Helper',
    settToken: TOKENS.BARB_SUSHI_WETH_SUSHI,
    depositToken: TOKENS.ARB_SUSHI_WETH_SUSHI,
    createdBlock: 13163959,
    protocol: Protocol.Sushiswap,
    strategy: '0x86f772C82914f5bFD168f99e208d0FC2C371e9C2',
  },
  {
    name: 'Sushiswap Wrapped BTC/Wrapped ETH',
    settToken: TOKENS.BARB_SUSHI_WETH_WBTC,
    depositToken: TOKENS.ARB_SUSHI_WETH_WBTC,
    createdBlock: 13163959,
    protocol: Protocol.Sushiswap,
    strategy: '0xA6827f0f14D0B83dB925B616d820434697328c22',
  },
  {
    name: 'Curve renBTC/wBTC',
    settToken: TOKENS.BARB_CRV_RENBTC,
    depositToken: TOKENS.ARB_CRV_RENBTC,
    getTokenBalance: getCurveSettTokenBalance,
    createdBlock: 13237551,
    protocol: Protocol.Curve,
  },
  {
    name: 'Curve Tricrypto',
    settToken: TOKENS.BARB_CRV_TRICRYPTO,
    depositToken: TOKENS.ARB_CRV_TRICRYPTO,
    getTokenBalance: getCurveSettTokenBalance,
    createdBlock: 13237551,
    protocol: Protocol.Curve,
  },
  {
    name: 'Curve Tricrypto Light',
    settToken: TOKENS.BARB_CRV_TRICRYPTO_LITE,
    depositToken: TOKENS.ARB_CRV_TRICRYPTO,
    getTokenBalance: getCurveSettTokenBalance,
    createdBlock: 13321375,
    stage: Stage.Staging,
    state: SettState.Experimental,
    protocol: Protocol.Curve,
  },
  {
    name: 'Swapr Swapr/Wrapped ETH',
    settToken: TOKENS.BARB_SWP_SWPR_WETH,
    depositToken: TOKENS.ARB_SWP_SWPR_WETH,
    createdBlock: 13315350,
    state: SettState.Guarded,
    protocol: Protocol.Swapr,
  },
  {
    name: 'Swapr Wrapped BTC/Wrapped ETH',
    settToken: TOKENS.BARB_SWP_WBTC_WETH,
    depositToken: TOKENS.ARB_SWP_WBTC_WETH,
    createdBlock: 13315350,
    state: SettState.Guarded,
    protocol: Protocol.Swapr,
  },
  {
    name: 'Swapr Badger/Wrapped ETH',
    bouncer: BouncerType.Badger,
    settToken: TOKENS.BARB_SWP_BADGER_WETH,
    depositToken: TOKENS.ARB_SWP_BADGER_WETH,
    createdBlock: 2188169,
    state: SettState.Guarded,
    protocol: Protocol.Swapr,
  },
];
