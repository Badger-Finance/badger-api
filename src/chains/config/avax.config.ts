import { Network, Protocol, VaultState, VaultVersion } from '@badger-dao/sdk';
import rpc from '../../config/rpc.config';
import { TOKENS } from '../../config/tokens.config';
import { GasPrices } from '../../gas/interfaces/gas-prices.interface';
import { VaultDefinition } from '../../vaults/interfaces/vault-definition.interface';
import { Chain } from './chain.config';
import { BaseStrategy } from '../strategies/base.strategy';
import { avalancheTokensConfig } from '../../tokens/config/avax-tokens.config';
import { Stage } from '../../config/enums/stage.enum';

export class Avalanche extends Chain {
  constructor() {
    super(
      'Avalanche',
      'avalanche',
      '0xa86a',
      Network.Avalanche,
      avalancheTokensConfig,
      rpc[Network.Avalanche],
      avalancheSetts,
      new BaseStrategy(Network.Avalanche, Object.keys(avalancheTokensConfig)),
    );
    Chain.register(this.network, this);
  }

  async getGasPrices(): Promise<GasPrices> {
    return this.defaultGasPrice();
  }

  // TODO: Update if badger is ever updated
  getBadgerTokenAddress(): string {
    return TOKENS.BADGER;
  }
}

export const avalancheSetts: VaultDefinition[] = [
  {
    name: 'WBTC',
    vaultToken: TOKENS.BAVAX_WBTC,
    depositToken: TOKENS.AVAX_WBTC,
    protocol: Protocol.Aave,
    stage: Stage.Staging,
    state: VaultState.Experimental,
    version: VaultVersion.v1_5,
  },
];
