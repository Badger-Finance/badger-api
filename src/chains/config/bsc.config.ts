import { Network } from "@badger-dao/sdk";

import rpc from "../../config/rpc.config";
import { TOKENS } from "../../config/tokens.config";
import { bscTokensConfig } from "../../tokens/config/bsc-tokens.config";
import { BaseStrategy } from "../strategies/base.strategy";
import { Chain } from "./chain.config";

export class BinanceSmartChain extends Chain {
  constructor() {
    super(Network.BinanceSmartChain, bscTokensConfig, rpc[Network.BinanceSmartChain], new BaseStrategy(Network.BinanceSmartChain));
    Chain.register(this.network, this);
  }

  getBadgerTokenAddress(): string {
    return TOKENS.MULTI_BADGER;
  }
}
