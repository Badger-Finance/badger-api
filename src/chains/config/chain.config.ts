import { BadRequest } from '@tsed/exceptions';
import { ethers } from 'ethers';
import { STAGE } from '../../config/constants';
import { GasPrices } from '../../gas/interfaces/gas-prices.interface';
import { SettDefinition } from '../../setts/interfaces/sett-definition.interface';
import { TokenConfig } from '../../tokens/interfaces/token-config.interface';
import { ChainStrategy } from '../strategies/chain.strategy';
import { Network } from '@badger-dao/sdk';

type Chains = Record<string, Chain>;

export abstract class Chain {
  private static chains: Chains = {};
  readonly name: string;
  readonly symbol: string;
  readonly chainId: string;
  readonly network: Network;
  readonly tokens: TokenConfig;
  readonly setts: SettDefinition[];
  readonly provider: ethers.providers.JsonRpcProvider;
  readonly batchProvider: ethers.providers.JsonRpcBatchProvider;
  readonly strategy: ChainStrategy;
  readonly graphUrl: string;
  readonly blocksPerYear: number;
  readonly badgerTree?: string;
  readonly rewardsLogger?: string;

  constructor(
    name: string,
    symbol: string,
    chainId: string,
    network: Network,
    tokens: TokenConfig,
    setts: SettDefinition[],
    rpcUrl: string,
    strategy: ChainStrategy,
    blocksPerYear: number,
    badgerTree?: string,
    rewardsLogger?: string,
  ) {
    this.name = name;
    this.symbol = symbol;
    this.chainId = chainId;
    this.network = network;
    this.tokens = tokens;
    this.setts = setts.filter((sett) => !sett.stage || sett.stage === STAGE);
    this.provider = new ethers.providers.JsonRpcProvider(rpcUrl);
    this.batchProvider = new ethers.providers.JsonRpcBatchProvider(rpcUrl);
    this.strategy = strategy;
    this.graphUrl = `https://api.thegraph.com/subgraphs/name/badger-finance/badger-dao-setts${
      network !== Network.Ethereum ? `-${symbol.toLowerCase()}` : ''
    }`;
    this.blocksPerYear = blocksPerYear;
    this.badgerTree = badgerTree;
    this.rewardsLogger = rewardsLogger;
  }

  static register(network: Network, chain: Chain): void {
    Chain.chains[network] = chain;
    Chain.chains[chain.symbol] = chain;
    if (network === Network.Polygon) {
      Chain.chains['matic'] = chain;
    }
  }

  static getChain(network?: Network): Chain {
    if (!network) {
      network = Network.Ethereum;
    }
    const chain = this.chains[network];
    if (!chain) {
      throw new BadRequest(`${network} is not a supported chain`);
    }
    return chain;
  }

  abstract getGasPrices(): Promise<GasPrices>;
}
