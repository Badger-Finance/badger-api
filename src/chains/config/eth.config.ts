import { ethers } from 'ethers';
import { BADGER_URL, Provider, STRATEGIES, TOKENS } from '../../config/constants';
import { Protocol } from '../../config/enums/protocol.enum';
import { Stage } from '../../config/enums/stage.enum';
import { getCurveSettTokenBalance } from '../../indexer/strategies/convex.strategy';
import { SettDefinition } from '../../setts/interfaces/sett-definition.interface';
import { ethTokensConfig } from '../../tokens/config/eth-tokens.config';
import { getZsDiggTokenBalance } from '../../tokens/custom/zs-digg-balance';
import { ChainNetwork } from '../enums/chain-network.enum';
import { EthStrategy } from '../strategies/eth.strategy';
import { Chain } from './chain.config';

export class Ethereum extends Chain {
  constructor() {
    super(
      'Ethereum',
      'eth',
      '0x01',
      ChainNetwork.Ethereum,
      ethTokensConfig,
      ethSetts,
      new ethers.providers.JsonRpcProvider(Provider.Alchemy),
      new EthStrategy(),
      BADGER_URL,
      2425847,
    );
    Chain.register(ChainNetwork.Ethereum, this);
  }
}

export const ethSetts: SettDefinition[] = [
  {
    name: 'Curve.fi renBTC/wBTC/sBTC',
    createdBlock: 11380877,
    depositToken: TOKENS.CRV_SBTC,
    getTokenBalance: getCurveSettTokenBalance,
    settToken: TOKENS.BCRV_SBTC,
    strategy: '0x8cbb86a7e0780a6fbefeec108f9b4b0aa8193e24',
    protocol: Protocol.Curve,
  },
  {
    name: 'Curve.fi renBTC/wBTC',
    createdBlock: 11380872,
    depositToken: TOKENS.CRV_RENBTC,
    getTokenBalance: getCurveSettTokenBalance,
    settToken: TOKENS.BCRV_RENBTC,
    strategy: '0xaa9b716ccd717761f40479cd81f8e3a5a7b4cad7',
    protocol: Protocol.Curve,
  },
  {
    name: 'Curve.fi tBTC/sBTC',
    createdBlock: 11380878,
    depositToken: TOKENS.CRV_TBTC,
    getTokenBalance: getCurveSettTokenBalance,
    settToken: TOKENS.BCRV_TBTC,
    strategy: '0x1ac31c470b90e366c70efc1ac28d5d7fa2f1dbe1',
    protocol: Protocol.Curve,
  },
  // TODO: Remove once fully deprecated
  {
    name: 'Harvest renBTC/wBTC',
    createdBlock: 11380939,
    depositToken: TOKENS.CRV_RENBTC,
    getTokenBalance: getCurveSettTokenBalance,
    settToken: TOKENS.BCRV_HRENBTC,
    protocol: Protocol.Curve,
  },
  {
    name: 'Uniswap Wrapped BTC/Badger',
    createdBlock: 11380883,
    depositToken: TOKENS.UNI_BADGER_WBTC,
    settToken: TOKENS.BUNI_BADGER_WBTC,
    protocol: Protocol.Uniswap,
  },
  {
    name: 'Uniswap Wrapped BTC/Digg',
    createdBlock: 11680833,
    depositToken: TOKENS.UNI_DIGG_WBTC,
    settToken: TOKENS.BUNI_DIGG_WBTC,
    protocol: Protocol.Uniswap,
  },
  {
    name: 'Sushiswap Wrapped BTC/Wrapped Ether',
    createdBlock: 11539529,
    depositToken: TOKENS.SUSHI_ETH_WBTC,
    strategy: '0x7A56d65254705B4Def63c68488C0182968C452ce',
    settToken: TOKENS.BSUSHI_ETH_WBTC,
    protocol: Protocol.Sushiswap,
  },
  {
    name: 'Sushiswap Wrapped BTC/Badger',
    createdBlock: 11537600,
    depositToken: TOKENS.SUSHI_BADGER_WBTC,
    strategy: '0x3a494D79AA78118795daad8AeFF5825C6c8dF7F1',
    settToken: TOKENS.BSUSHI_BADGER_WBTC,
    protocol: Protocol.Sushiswap,
  },
  {
    name: 'Sushiswap Wrapped BTC/Digg',
    createdBlock: 11681030,
    depositToken: TOKENS.SUSHI_DIGG_WBTC,
    strategy: '0xaa8dddfe7DFA3C3269f1910d89E4413dD006D08a',
    settToken: TOKENS.BSUSHI_DIGG_WBTC,
    protocol: Protocol.Sushiswap,
  },
  {
    name: 'Digg',
    createdBlock: 11680422,
    depositToken: TOKENS.DIGG,
    settToken: TOKENS.BDIGG,
  },
  {
    name: 'Badger',
    createdBlock: 11380871,
    depositToken: TOKENS.BADGER,
    settToken: TOKENS.BBADGER,
  },
  {
    createdBlock: 12246710,
    name: 'Yearn Wrapped BTC',
    depositToken: TOKENS.WBTC,
    settToken: TOKENS.BVYWBTC,
  },
  {
    name: 'Sushiswap Wrapped BTC/ibBTC',
    createdBlock: 12383984,
    depositToken: TOKENS.SUSHI_IBBTC_WBTC,
    settToken: TOKENS.BSUSHI_IBBTC_WBTC,
    strategy: '0xf4146A176b09C664978e03d28d07Db4431525dAd',
    protocol: Protocol.Sushiswap,
  },
  {
    name: 'Digg Stabilization',
    balanceDecimals: 18,
    createdBlock: 12375910,
    depositToken: TOKENS.DIGG,
    experimental: true,
    getTokenBalance: getZsDiggTokenBalance,
    protocol: Protocol.Badger,
    settToken: TOKENS.BZS_DIGG,
    stage: Stage.Staging,
    strategy: STRATEGIES.BZS_DIGG,
  },
  {
    name: 'Convex hBTC',
    createdBlock: 12589485,
    depositToken: TOKENS.CRV_HBTC,
    experimental: true,
    getTokenBalance: getCurveSettTokenBalance,
    settToken: TOKENS.BCRV_HBTC,
    strategy: '0xff26f400e57bf726822eacbb64fa1c52f1f27988',
    protocol: Protocol.Convex,
  },
  {
    name: 'Convex pBTC',
    createdBlock: 12589485,
    depositToken: TOKENS.CRV_PBTC,
    experimental: true,
    getTokenBalance: getCurveSettTokenBalance,
    settToken: TOKENS.BCRV_PBTC,
    strategy: '0x1c1fd689103bbfd701b3b7d41a3807f12814033d',
    protocol: Protocol.Convex,
  },
  {
    name: 'Convex oBTC',
    createdBlock: 12589485,
    depositToken: TOKENS.CRV_OBTC,
    experimental: true,
    getTokenBalance: getCurveSettTokenBalance,
    settToken: TOKENS.BCRV_OBTC,
    strategy: '0x2bb864cdb4856ab2d148c5ca52dd7ccec126d138',
    protocol: Protocol.Convex,
  },
  {
    name: 'Convex bBTC',
    createdBlock: 12589485,
    depositToken: TOKENS.CRV_BBTC,
    experimental: true,
    getTokenBalance: getCurveSettTokenBalance,
    settToken: TOKENS.BCRV_BBTC,
    strategy: '0x4f3e7a4566320b2709fd1986f2e9f84053d3e2a0',
    protocol: Protocol.Convex,
  },
  {
    name: 'Convex Tricrypto',
    createdBlock: 12679976,
    depositToken: TOKENS.CRV_TRICRYPTO,
    experimental: true,
    getTokenBalance: getCurveSettTokenBalance,
    settToken: TOKENS.BCRV_TRICRYPTO,
    strategy: '0x05ec4356e1acd89cc2d16adc7415c8c95e736ac1',
    protocol: Protocol.Convex,
  },
  {
    name: 'Convex cvxCRV Helper',
    createdBlock: 12678302,
    depositToken: TOKENS.CVXCRV,
    experimental: true,
    settToken: TOKENS.BCVXCRV,
    strategy: '0x4FE836b28ADc7867290764885b24d3992dAeE819',
    protocol: Protocol.Convex,
  },
  {
    name: 'Convex CVX Helper',
    createdBlock: 12678302,
    depositToken: TOKENS.CVX,
    experimental: true,
    settToken: TOKENS.BCVX,
    strategy: '0x5e7a939f77E5c441e2Ba1Dc8Be5D42020193376D',
    protocol: Protocol.Convex,
  },
];
