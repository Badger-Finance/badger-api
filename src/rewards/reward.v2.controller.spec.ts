import BadgerSDK, { RewardsService } from '@badger-dao/sdk';
import rewardsLoadSchedulesMock from '@badger-dao/sdk-mocks/generated/ethereum/rewards/loadSchedules.json';
import { PlatformTest } from '@tsed/common';
import { BadRequest, NotFound } from '@tsed/exceptions';
import SuperTest from 'supertest';

import { TOKENS } from '../config/tokens.config';
import { Server } from '../Server';
import { mockChainVaults } from '../test/tests.utils';

describe('RewardController', () => {
  let request: SuperTest.SuperTest<SuperTest.Test>;

  const schedulesMockMap = {
    [TOKENS.BBADGER]: rewardsLoadSchedulesMock,
    [TOKENS.BCVX]: rewardsLoadSchedulesMock.map((rw) => ({
      ...rw,
      beneficiary: TOKENS.BCVX,
    })),
  };

  const activeSchedulesMockMap = {
    [TOKENS.BBADGER]: schedulesMockMap[TOKENS.BBADGER].map((rw) => ({
      ...rw,
      compPercent: 50,
    })),
    [TOKENS.BCVX]: schedulesMockMap[TOKENS.BCVX].map((rw) => ({
      ...rw,
      compPercent: 70,
    })),
  };

  function setupDefaultMocks() {
    jest.spyOn(RewardsService.prototype, 'loadSchedules').mockImplementation(async (beneficiary) => {
      return schedulesMockMap[beneficiary];
    });
    jest.spyOn(RewardsService.prototype, 'loadActiveSchedules').mockImplementation(async (beneficiary) => {
      return activeSchedulesMockMap[beneficiary];
    });
    jest.spyOn(BadgerSDK.prototype, 'ready').mockImplementation();
    mockChainVaults();
  }

  beforeAll(PlatformTest.bootstrap(Server));
  beforeAll(async () => {
    request = SuperTest(PlatformTest.callback());
  });
  afterAll(PlatformTest.reset);

  describe('GET /v2/reward/schedules', () => {
    describe('with no specified chain', () => {
      it('returns schedules for default chain and all vaults', async (done: jest.DoneCallback) => {
        setupDefaultMocks();
        const { body } = await request.get('/v2/reward/schedules').expect(200);
        expect(body).toMatchSnapshot();
        done();
      });
    });

    describe('with active param true', () => {
      it('returns active schedules for default chain and all vaults', async (done: jest.DoneCallback) => {
        setupDefaultMocks();
        const { body } = await request.get('/v2/reward/schedules?active=true').expect(200);
        expect(body).toMatchSnapshot();
        done();
      });
    });

    describe('with an invalid specified chain', () => {
      it('returns a 400', async (done: jest.DoneCallback) => {
        setupDefaultMocks();
        const { body } = await request.get('/v2/reward/schedules?chain=invalid').expect(BadRequest.STATUS);
        expect(body).toMatchSnapshot();
        done();
      });
    });
  });

  describe('GET /v2/reward/schedules/<beneficiary>', () => {
    describe('with no specified chain', () => {
      it('returns schedule for default chain and one vault', async (done: jest.DoneCallback) => {
        setupDefaultMocks();
        const { body } = await request.get(`/v2/reward/schedules/${TOKENS.BBADGER}`).expect(200);
        expect(body).toMatchSnapshot();
        done();
      });
    });

    describe('with active param true', () => {
      it('returns schedules for default chain and one vault', async (done: jest.DoneCallback) => {
        setupDefaultMocks();
        const { body } = await request.get(`/v2/reward/schedules/${TOKENS.BBADGER}?active=true`).expect(200);
        expect(body).toMatchSnapshot();
        done();
      });
    });

    describe('with an invalid specified chain', () => {
      it('returns a 400', async (done: jest.DoneCallback) => {
        setupDefaultMocks();
        const { body } = await request
          .get(`/v2/reward/schedules/${TOKENS.BBADGER}?chain=invalid`)
          .expect(BadRequest.STATUS);
        expect(body).toMatchSnapshot();
        done();
      });
    });

    describe('with invalid param specified', () => {
      it('returns a 400, NotFound', async (done: jest.DoneCallback) => {
        setupDefaultMocks();
        const { body } = await request.get(`/v2/reward/schedules/unknowsvaultdata`).expect(NotFound.STATUS);
        expect(body).toMatchSnapshot();
        done();
      });
    });
  });
});
