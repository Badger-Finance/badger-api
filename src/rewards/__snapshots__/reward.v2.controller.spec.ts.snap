// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reward.v2.controller GET /v2/reward/schedules with active param true returns active schedules for default chain and all vaults 1`] = `
Object {
  "0x19d97d8fa813ee2f51ad4b4e04ea08baf4dffc28": Array [
    Object {
      "amount": 2035.82,
      "beneficiary": "0x19d97d8fa813ee2f51ad4b4e04ea08baf4dffc28",
      "compPercent": 50,
      "end": 1619715600,
      "start": 1619110800,
      "token": "0x3472A5A71965499acd81997a54BBA8D852C6E53d",
    },
    Object {
      "amount": 0.094819176,
      "beneficiary": "0x19d97d8fa813ee2f51ad4b4e04ea08baf4dffc28",
      "compPercent": 50,
      "end": 1617901200,
      "start": 1617296400,
      "token": "0x798D1bE841a82a273720CE31c822C61a67a601C3",
    },
    Object {
      "amount": 1865.215,
      "beneficiary": "0x19d97d8fa813ee2f51ad4b4e04ea08baf4dffc28",
      "compPercent": 50,
      "end": 1620320400,
      "start": 1619715600,
      "token": "0x3472A5A71965499acd81997a54BBA8D852C6E53d",
    },
  ],
}
`;

exports[`reward.v2.controller GET /v2/reward/schedules with an invalid specified chain returns a 400 1`] = `
Object {
  "message": "invalid is not a supported chain",
}
`;

exports[`reward.v2.controller GET /v2/reward/schedules with no specified chain returns schedules for default chain and all vaults 1`] = `
Object {
  "0x19d97d8fa813ee2f51ad4b4e04ea08baf4dffc28": Array [
    Object {
      "amount": 2035.82,
      "beneficiary": "0x19d97d8fa813ee2f51ad4b4e04ea08baf4dffc28",
      "compPercent": 100,
      "end": 1619715600,
      "start": 1619110800,
      "token": "0x3472A5A71965499acd81997a54BBA8D852C6E53d",
    },
    Object {
      "amount": 0.094819176,
      "beneficiary": "0x19d97d8fa813ee2f51ad4b4e04ea08baf4dffc28",
      "compPercent": 100,
      "end": 1617901200,
      "start": 1617296400,
      "token": "0x798D1bE841a82a273720CE31c822C61a67a601C3",
    },
    Object {
      "amount": 1865.215,
      "beneficiary": "0x19d97d8fa813ee2f51ad4b4e04ea08baf4dffc28",
      "compPercent": 100,
      "end": 1620320400,
      "start": 1619715600,
      "token": "0x3472A5A71965499acd81997a54BBA8D852C6E53d",
    },
  ],
}
`;

exports[`reward.v2.controller GET /v2/reward/schedules/<beneficiary> with active param true returns schedules for default chain and one vault 1`] = `
Array [
  Object {
    "amount": 2035.82,
    "beneficiary": "0x19d97d8fa813ee2f51ad4b4e04ea08baf4dffc28",
    "compPercent": 50,
    "end": 1619715600,
    "start": 1619110800,
    "token": "0x3472A5A71965499acd81997a54BBA8D852C6E53d",
  },
  Object {
    "amount": 0.094819176,
    "beneficiary": "0x19d97d8fa813ee2f51ad4b4e04ea08baf4dffc28",
    "compPercent": 50,
    "end": 1617901200,
    "start": 1617296400,
    "token": "0x798D1bE841a82a273720CE31c822C61a67a601C3",
  },
  Object {
    "amount": 1865.215,
    "beneficiary": "0x19d97d8fa813ee2f51ad4b4e04ea08baf4dffc28",
    "compPercent": 50,
    "end": 1620320400,
    "start": 1619715600,
    "token": "0x3472A5A71965499acd81997a54BBA8D852C6E53d",
  },
]
`;

exports[`reward.v2.controller GET /v2/reward/schedules/<beneficiary> with an invalid specified chain returns a 400 1`] = `
Object {
  "message": "invalid is not a supported chain",
}
`;

exports[`reward.v2.controller GET /v2/reward/schedules/<beneficiary> with invalid param specified returns a 400, NotFound 1`] = `
Object {
  "code": 2002,
  "message": "Vault unknowsvaultdata doest have any data",
}
`;

exports[`reward.v2.controller GET /v2/reward/schedules/<beneficiary> with no specified chain returns schedule for default chain and one vault 1`] = `
Array [
  Object {
    "amount": 2035.82,
    "beneficiary": "0x19d97d8fa813ee2f51ad4b4e04ea08baf4dffc28",
    "compPercent": 100,
    "end": 1619715600,
    "start": 1619110800,
    "token": "0x3472A5A71965499acd81997a54BBA8D852C6E53d",
  },
  Object {
    "amount": 0.094819176,
    "beneficiary": "0x19d97d8fa813ee2f51ad4b4e04ea08baf4dffc28",
    "compPercent": 100,
    "end": 1617901200,
    "start": 1617296400,
    "token": "0x798D1bE841a82a273720CE31c822C61a67a601C3",
  },
  Object {
    "amount": 1865.215,
    "beneficiary": "0x19d97d8fa813ee2f51ad4b4e04ea08baf4dffc28",
    "compPercent": 100,
    "end": 1620320400,
    "start": 1619715600,
    "token": "0x3472A5A71965499acd81997a54BBA8D852C6E53d",
  },
]
`;
